In [1]:
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


In [2]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt

In [3]:
#input MNIST
from keras.datasets import mnist

Using TensorFlow backend.


In [4]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

### 輸入格式整理

In [5]:
# CNN 要的是 (28,28,1)
x_train = x_train.reshape(60000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

### 輸出格式整理

In [6]:
from keras.utils import np_utils

In [7]:
y_train = np_utils.to_categorical(y_train,10)
y_test = np_utils.to_categorical(y_test, 10)

In [8]:
#normalize
x_train = x_train/255
x_test = x_test/255

### 建立CNN

In [9]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

In [10]:
model_1 = Sequential()

In [11]:
#filter為16, Kernal size 為 5x5, padding為'same', activation為relu
model_1.add(Conv2D(16, (5,5), padding='same', input_shape=(28,28,1)))
model_1.add(Activation('relu'))

Instructions for updating:
Colocations handled automatically by placer.


In [12]:
#MaxPooling size 為 (2,2)
model_1.add(MaxPool2D(pool_size=(2,2)))

In [14]:
model_1.add(Conv2D(36, (5,5), padding='same'))
model_1.add(Activation('relu'))

In [15]:
model_1.add(MaxPool2D(pool_size=(2,2)))

In [16]:
model_1.add(Conv2D(49, (5,5), padding='same'))
model_1.add(Activation('relu'))

In [18]:
model_1.add(MaxPool2D(pool_size=(2,2)))

In [20]:
model_1.add(Flatten())
model_1.add(Dense(9))
model_1.add(Activation('relu'))

In [21]:
model_1.add(Dense(10))
model_1.add(Activation('softmax'))

In [23]:
model_1.compile(loss='categorical_crossentropy', optimizer=SGD(lr=0.07), metrics=['accuracy'])

In [24]:
model_1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 28, 28, 16)        416       
_________________________________________________________________
activation_1 (Activation)    (None, 28, 28, 16)        0         
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 16)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 14, 14, 36)        14436     
_________________________________________________________________
activation_2 (Activation)    (None, 14, 14, 36)        0         
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 7, 7, 36)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 7, 7, 49)          44149     
__________

In [25]:
model_1.fit(x_train, y_train, batch_size=100, epochs=10)

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
60000/60000 [==============================] - 78s 1ms/step - loss: 0.4793 - acc: 0.8389
Epoch 2/10
60000/60000 [==============================] - 78s 1ms/step - loss: 0.0871 - acc: 0.9727
Epoch 3/10
60000/60000 [==============================] - 77s 1ms/step - loss: 0.0571 - acc: 0.9824
Epoch 4/10
60000/60000 [==============================] - 77s 1ms/step - loss: 0.0449 - acc: 0.9859
Epoch 5/10
60000/60000 [==============================] - 76s 1ms/step - loss: 0.0354 - acc: 0.9889
Epoch 6/10
60000/60000 [==============================] - 76s 1ms/step - loss: 0.0292 - acc: 0.9911
Epoch 7/10
60000/60000 [==============================] - 77s 1ms/step - loss: 0.0241 - acc: 0.9926
Epoch 8/10
60000/60000 [==============================] - 80s 1ms/step - loss: 0.0209 - acc: 0.9936
Epoch 9/10
60000/60000 [==============================] - 78s 1ms/step - loss: 0.0178 - acc: 0.9943
Epoch 10/10
60000/60000 [===========================

In [26]:
score = model_1.evaluate(x_test, y_test)
print('loss:', score[0])
print('accuracy:', score[1])

10000/10000 [==============================] - 5s 534us/step
loss: 0.03142442755378943
accuracy: 0.9902


In [11]:
model_2 = Sequential()

In [21]:
#filter為16, Kernal size 為 5x5, padding為'same', activation為relu
model_2.add(Conv2D(16, (5,5), padding='same', input_shape=(28,28,1)))
model_2.add(Activation('relu'))

In [22]:
#MaxPooling size 為 (3,3)
model_2.add(MaxPool2D(pool_size=(3,3)))

In [23]:
model_2.add(Conv2D(36, (5,5), padding='same'))
model_2.add(Activation('relu'))

In [24]:
model_2.add(MaxPool2D(pool_size=(3,3)))

In [25]:
model_2.add(Conv2D(49, (5,5), padding='same'))
model_1.add(Activation('relu'))

In [26]:
model_2.add(MaxPool2D(pool_size=(3,3)))

In [27]:
model_2.add(Flatten())
model_2.add(Dense(9))
model_2.add(Activation('relu'))

In [28]:
model_2.add(Dense(10))
model_2.add(Activation('softmax'))

In [29]:
model_2.compile(loss='categorical_crossentropy', optimizer=SGD(lr=0.07), metrics=['accuracy'])

In [30]:
model_2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 28, 28, 16)        416       
_________________________________________________________________
activation_6 (Activation)    (None, 28, 28, 16)        0         
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 9, 9, 16)          0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 9, 9, 36)          14436     
_________________________________________________________________
activation_7 (Activation)    (None, 9, 9, 36)          0         
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 3, 3, 36)          0         
_________________________________________________________________
conv2d_6 (Conv2D)            (None, 3, 3, 49)          44149     
__________

In [31]:
model_2.fit(x_train, y_train, batch_size=100, epochs=10)

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
60000/60000 [==============================] - 43s 722us/step - loss: 0.6523 - acc: 0.7855
Epoch 2/10
60000/60000 [==============================] - 43s 710us/step - loss: 0.1032 - acc: 0.9693
Epoch 3/10
60000/60000 [==============================] - 44s 727us/step - loss: 0.0681 - acc: 0.9787
Epoch 4/10
60000/60000 [==============================] - 42s 708us/step - loss: 0.0515 - acc: 0.9840
Epoch 5/10
60000/60000 [==============================] - 43s 715us/step - loss: 0.0410 - acc: 0.9875
Epoch 6/10
60000/60000 [==============================] - 43s 712us/step - loss: 0.0344 - acc: 0.9887
Epoch 7/10
60000/60000 [==============================] - 43s 725us/step - loss: 0.0284 - acc: 0.9906
Epoch 8/10
60000/60000 [==============================] - 43s 720us/step - loss: 0.0254 - acc: 0.9921
Epoch 9/10
60000/60000 [==============================] - 42s 706us/step - loss: 0.0209 - acc: 0.9934
Epoch 10/10
60000/60000 [=========

In [32]:
score = model_2.evaluate(x_test, y_test)
print('loss:', score[0])
print('accuracy:', score[1])

10000/10000 [==============================] - 3s 335us/step
loss: 0.03691538950659451
accuracy: 0.9885


In [33]:
model_3 = Sequential()

In [35]:
#filter為16, Kernal size 為 5x5, padding為'same', activation為relu
model_3.add(Conv2D(16, (5,5), padding='same', input_shape=(28,28,1)))
model_3.add(Activation('relu'))

In [36]:
#MaxPooling size 為 (2,2)
model_3.add(MaxPool2D(pool_size=(2,2)))

In [37]:
model_3.add(Conv2D(36, (5,5), padding='same'))
model_3.add(Activation('relu'))

In [38]:
model_3.add(MaxPool2D(pool_size=(2,2)))

In [39]:
model_3.add(Conv2D(49, (5,5), padding='same'))
model_3.add(Activation('relu'))

In [40]:
model_3.add(MaxPool2D(pool_size=(2,2)))

In [41]:
model_3.add(Flatten())
model_3.add(Dense(9))
model_3.add(Activation('relu'))

In [42]:
model_3.add(Dense(10))
model_3.add(Activation('softmax'))

In [44]:
model_3.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [45]:
model_3.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_8 (Conv2D)            (None, 28, 28, 16)        416       
_________________________________________________________________
activation_11 (Activation)   (None, 28, 28, 16)        0         
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 14, 14, 16)        0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 14, 14, 36)        14436     
_________________________________________________________________
activation_12 (Activation)   (None, 14, 14, 36)        0         
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 7, 7, 36)          0         
_________________________________________________________________
conv2d_10 (Conv2D)           (None, 7, 7, 49)          44149     
__________

In [46]:
model_3.fit(x_train, y_train, batch_size=100, epochs=10)

Epoch 1/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.3377 - acc: 0.8961
Epoch 2/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.0785 - acc: 0.9765
Epoch 3/10
60000/60000 [==============================] - 81s 1ms/step - loss: 0.0511 - acc: 0.9847
Epoch 4/10
60000/60000 [==============================] - 81s 1ms/step - loss: 0.0389 - acc: 0.9888
Epoch 5/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.0306 - acc: 0.9911
Epoch 6/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.0248 - acc: 0.9922
Epoch 7/10
60000/60000 [==============================] - 81s 1ms/step - loss: 0.0202 - acc: 0.9934
Epoch 8/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.0171 - acc: 0.9944
Epoch 9/10
60000/60000 [==============================] - 82s 1ms/step - loss: 0.0144 - acc: 0.9953
Epoch 10/10
60000/60000 [==============================] - 79s 1ms/step - loss: 0.0121 - acc: 0.9960

In [47]:
score = model_3.evaluate(x_test, y_test)
print('loss:', score[0])
print('accuracy:', score[1])

10000/10000 [==============================] - 5s 539us/step
loss: 0.03594355874414068
accuracy: 0.9901


### 小結

* 延續課堂上的例子，model_1 將filter的大小加大，testing accuracy 從0.86上升到0.99
* model_2 則是延續 model_1，增加 max pooling 的大小，testing accuracy 反而下降成0.98
* 最後一個 model_3 則是延續 model_1 只是改變 optimizer 為 adam，testing accuracy 沒有變